pandas/core/tools/datetimes.py

```diff
@@ -14,12 +15,25 @@
 from pandas.core.dtypes.common import (
     ensure_object, is_datetime64_dtype, is_datetime64_ns_dtype,
     is_datetime64tz_dtype, is_float, is_integer, is_integer_dtype,
-    is_list_like, is_numeric_dtype, is_object_dtype, is_scalar)
-from pandas.core.dtypes.generic import ABCDataFrame, ABCIndexClass, ABCSeries
+    is_list_like, is_numeric_dtype, is_scalar)
+from pandas.core.dtypes.generic import (
+    ABCDataFrame, ABCDatetimeIndex, ABCIndex, ABCIndexClass, ABCSeries)
 from pandas.core.dtypes.missing import notna
 
+from pandas._typing import ArrayLike
 from pandas.core import algorithms
 
+# ---------------------------------------------------------------------
+# types used in annotations
+
+Scalar = Union[int, float, str]
+DatetimeScalar = Union[Scalar, datetime]
```

Comment:
@WillAyd look ok to you?

Reply:
@anmyachev we already have DatetimeLikeScalar defined in pandas._typing - should be able to import from there

w.r.t Scalar probably better in pandas._typing as well but can clean that up separate to this PR so it's not a blocker

Reply:
these are different (much less strict) as. regular DatetimeScalar

Reply:
Ah OK makes sense. Probably better as a TypeVar for this module. Can clean up as follow up or @anmyachev if you wanted to do here OK as well.

Would look something like `DatetimeScalar = TypeVar('DatetimeScalar', Scalar, datetime)`

Reply:
I could do it only for `DatetimeScalar`

Reply:
@WillAyd If I try to do something like that: 
```
Scalar = TypeVar('Scalar', int, float, str)
DatetimeScalar = TypeVar('DatetimeScalar', Scalar, datetime)
```
then mypy will find the next problem:
`pandas\core\tools\datetimes.py:30: error: Invalid type "pandas.core.tools.datetimes.Scalar"`

Reply:
Yea no problem. Scalar belongs in pandas._typing anyway and I'm not sure if that itself should be a TypeVar (we have a lot of cases where int and float are used interchangeably within a function) so let's leave that to a separate PR to clean up

Reply:
Appreciate you taking a look though!